### DISCUSSION TITLE
Deleting open file doesn't create new buffer

### DISCUSSION THREAD

**Discussion by Raagh** [nvim-tree/nvim-tree.lua#2034] [Q&A]
### Description

Hi , I have weird case that maybe you can help me fix.

I am using toggleterm and nvim-tree together.

![nvim-2023-02-21-12-03-58](https://user-images.githubusercontent.com/8405459/220331889-3a3dcb8c-5953-40ac-b9fb-45b50da69a9f.gif)

### Neovim version

```text
nvim 0.8.2
```

### Operating system and version

Windows 11(WSL Ubuntu)

### nvim-tree version

latest

### Minimal config

```lua
-- each of these are documented in `:help nvim-tree.OPTION_NAME`
local status_ok, nvim_tree = pcall(require, "nvim-tree")
if not status_ok then
  return
end

local config_status_ok, _ = pcall(require, "nvim-tree.config")
if not config_status_ok then
  return
end

nvim_tree.setup {
  disable_netrw = true,
  hijack_netrw = true,
  update_cwd = true,
  update_focused_file = {
    enable = true,
    update_cwd = true,
    ignore_list = {}
  },
  trash = {
    cmd = "trash",
    require_confirm = true,
  },
  actions = {
    open_file = {
      resize_window = false,
    }
  },
  view = {
    adaptive_size = true,
    mappings = {
      list = {
        { key = "e",  action = "" },
        { key = "<S-r>",  action = "rename_basename" },
      }
    }
  }
}

vim.api.nvim_create_autocmd('BufEnter', {
  command = "if winnr('$') == 1 && bufname() == 'NvimTree_' . tabpagenr() | quit | endif",
  nested = true,
})


local events, nvim_tree_events = pcall(require, "nvim-tree.events")
if not events then
  return
end

local bufferapi, bufferline_api = pcall(require, "bufferline.api")
if not bufferapi then
  return
end

local function get_tree_size()
  return require 'nvim-tree.view'.View.width
end

nvim_tree_events.subscribe('TreeOpen', function()
  bufferline_api.set_offset(get_tree_size())
end)

nvim_tree_events.subscribe('Resize', function()
  bufferline_api.set_offset(get_tree_size())
end)

nvim_tree_events.subscribe('TreeClose', function()
  bufferline_api.set_offset(0)
end)
```

### Steps to reproduce

    Open a file
    Open toggleterm as horizontal
    Open nvim-tree
    Delete file which is open

### Expected behavior

    nvim-tree would create a new empty buffer if there is no other buffer open

### Actual behavior

   nvim-tree leaves no buffer open

**Comment by gegoune** [comment]
This isn't as much about buffers as it is about windows. When deleting file buffer will get removed together with window it's displayed in.

**Comment by Raagh** [comment]
Sorry, I am kind of new about neovim so I probably explain it wrong. To clarify I am not saying this is a bug for nvim-tree, might have to solve this from toggleterm but I wonder if there is a way to fix this from this plugin

**Comment by gegoune** [comment]
Nothing to be sorry about. It is bit annoying indeed, every time I delete file its window will be removed as well making nvim-tree fullscreen. Maybe @alex-courtis will have an idea how to handle this. We could open empty buffer in new window I guess.

Also I don't think it has anything to do with toggleterm. You most likely going to get the same behaviour without it.

**Comment by charbelnicolas** [comment]
I'm also getting a full screen nvim tree window when closing a buffer even when there are still more buffers to show. Any ideas of why this is happening?

**Comment by gegoune** [comment]
I believe this is just default vim behaviour. It doesn't matter what those windows are. Close one window and others will expand to fit regained space. If only another one was left it will end up taking entire "window". 

I am bit afraid that difficulty of "quitting when nvim-tree is the only window left" will also apply to scenario when it would try to open new window with either last opened buffer if present or new no name buffer.

Is aforementioned case the same as 'quit-if-last', @alex-courtis, or do you think it could be reliably implemented?

**Comment by alex-courtis** [comment]
> Nothing to be sorry about. It is bit annoying indeed, every time I delete file its window will be removed as well making nvim-tree fullscreen. Maybe @alex-courtis will have an idea how to handle this. We could open empty buffer in new window I guess.

This is annoying behaviour, but it's just how vim works.

There is a plugin to work around it: https://github.com/qpkorr/vim-bufkill

`:BD` and friends will not close the window when a buffer is deleted.

**Comment by alex-courtis** [comment]
> I am bit afraid that difficulty of "quitting when nvim-tree is the only window left" will also apply to scenario when it would try to open new window with either last opened buffer if present or new no name buffer.
> 
> Is aforementioned case the same as 'quit-if-last', @alex-courtis, or do you think it could be reliably implemented?

That is indeed the same issue. vim events are only somewhat deterministic, in that other plugins and automation will change order or suppress events.

I don't think this is something we could introduce as it will likely introduce a _lot_ of bugs.

**Comment by charbelnicolas** [comment]
https://github.com/famiu/bufdelete.nvim

This one works too!

**Comment by alex-courtis** [comment]
> https://github.com/famiu/bufdelete.nvim
> 
> This one works too!

Today I learned.

It's quite nice, however I can't move as I'm using the MRU features of bufkill. Not sure why they are not in a separate plugin...
